packages/panzer/disc-fe/src/Panzer_L2Projection.cpp

```diff
@@ -151,106 +161,90 @@ namespace panzer {
         }
         const auto& basisValues = *bv_ptr;
 
-        // Leaving the commented out workset loop in teh code for now
-        // in case we want to support the original workset
-        // construction path in the workset container that is used
-        // here.
+        const auto unweightedBasis = basisValues.getBasisValues(false).get_static_view();
+        const auto weightedBasis = basisValues.getBasisValues(true).get_static_view();
 
-        // Based on descriptor, currently assumes there should only be one workset
-        // panzer::WorksetDescriptor wd(block,panzer::WorksetSizeType::ALL_ELEMENTS,true,true);
-        // const auto worksets = worksetContainer_->getWorksets(wd);
+        const std::vector<panzer::LocalOrdinal>& offsets = targetGlobalIndexer_->getGIDFieldOffsets(block,fieldIndex);
+        PHX::View<panzer::LocalOrdinal*> kOffsets("MassMatrix: Offsets",offsets.size());
+        auto kOffsets_h = Kokkos::create_mirror_view(kOffsets);
 
-        // for (const auto& workset : *worksets) {
+        for(const auto& i : offsets)
+          kOffsets_h(i) = offsets[i];
 
-        //   const auto basisValues = workset.getBasisValues(targetBasisDescriptor_,integrationDescriptor_);
+        Kokkos::deep_copy(kOffsets, kOffsets_h);
 
-          const auto unweightedBasis = basisValues.getBasisValues(false).get_static_view();
-          const auto weightedBasis = basisValues.getBasisValues(true).get_static_view();
+        // Local Ids
+        PHX::View<panzer::LocalOrdinal**> localIds("MassMatrix: LocalIds", num_cells_owned_ghosted_virtual,
+                                                   targetGlobalIndexer_->getElementBlockGIDCount(block));
 
-          const std::vector<panzer::LocalOrdinal>& offsets = targetGlobalIndexer_->getGIDFieldOffsets(block,fieldIndex);
-          PHX::View<panzer::LocalOrdinal*> kOffsets("MassMatrix: Offsets",offsets.size());
-          auto kOffsets_h = Kokkos::create_mirror_view(kOffsets);
+        // Remove the ghosted cell ids or the call to getElementLocalIds will spill array bounds
+        const auto cellLocalIdsNoGhost = Kokkos::subview(cell_local_ids,std::make_pair(0,num_cells_owned));
+
+        targetGlobalIndexer_->getElementLIDs(cellLocalIdsNoGhost,localIds);
 
-          for(const auto& i : offsets)
-            kOffsets_h(i) = offsets[i];
+        const int numBasisPoints = static_cast<int>(weightedBasis.extent(1));
+        if ( use_lumping ) {
+          Kokkos::parallel_for(num_cells_owned,KOKKOS_LAMBDA (const int& cell) {
+            double total_mass = 0.0, trace = 0.0;
 
-          Kokkos::deep_copy(kOffsets, kOffsets_h);
+            panzer::LocalOrdinal cLIDs[256];
```

Comment:
Define the magic 256 via macro?

Reply:
I can change that. Thanks! That particular code has been in panzer for years. I just fixed the indentation as an outer loop was removed. Might be easier to review by looking at the different commits (if its not too late). The last commit just cleaned up indentation. Functional changes were first two commits.

Reply:
Sounds good.